### PR TITLE
Drop ruby 1.8 support and use IO.popen

### PR DESCRIPTION
**Pull request by nobu**
IO.popen supports command line as an array since 1.9, as well as
system had done.  And both support redirection option since 1.9
too.  These prevents from excution by shell which could cause
unexpected expansions.

# Description:

______________

# Tasks:

- [x] Describe the problem / feature
- [ ] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems/ext/builder.rb [hunk 1]
```diff
@@ -56,23 +56,23 @@ def self.make(dest_path, results)
...
       rubygems_gemdeps, ENV['RUBYGEMS_GEMDEPS'] = ENV['RUBYGEMS_GEMDEPS'], nil
       if verbose
         puts("current directory: #{Dir.pwd}")
         puts(command)
         system(command)
       else
         results << "current directory: #{Dir.pwd}"
-        results << command
-        results << `#{command} #{redirector}`
+        results << (Array === command ? command.shelljoin : command)
```
Comment (duckinator) [inline lib/rubygems/ext/builder.rb]: Is there any reason to prefer `Array === command` over `command.is_a?(Array)`? If not, could you switch to the latter?

File: lib/rubygems/ext/rake_builder.rb [hunk 2]
```diff
@@ -18,15 +18,18 @@ def self.build(extension, dest_path, results, args=[], lib_dir=nil)
...
+      begin
+        rake = [Gem.ruby, "-rrubygems", Gem.bin_path('rake', 'rake')]
+      rescue Gem::Exception
+        rake = [Gem.default_exec_format % 'rake']
+      end
+    end
 
     rake_args = ["RUBYARCHDIR=#{dest_path}", "RUBYLIBDIR=#{dest_path}", *args]
-    run "#{rake} #{rake_args.shelljoin}", results
+    run rake+rake_args, results
```
Comment (duckinator) [inline lib/rubygems/ext/rake_builder.rb]: Could you change this to `run(rake + rake_args, results)`?

File: lib/rubygems/ext/rake_builder.rb [hunk 3]
```diff
@@ -18,15 +18,18 @@ def self.build(extension, dest_path, results, args=[], lib_dir=nil)
...
-    rake ||= Gem.default_exec_format % 'rake'
+    if rake
+      rake = rake.shellsplit
+    else
+      begin
+        rake = [Gem.ruby, "-rrubygems", Gem.bin_path('rake', 'rake')]
+      rescue Gem::Exception
+        rake = [Gem.default_exec_format % 'rake']
+      end
+    end
```
Comment (duckinator) [inline lib/rubygems/ext/rake_builder.rb]: Very nice! This chunk of code is much nicer than what we had before. :+1:

File: lib/rubygems/test_case.rb [hunk 4]
```diff
@@ -1319,9 +1319,18 @@ def self.rubybin
...
+      "\"#{path.gsub(/[`$"]/, '\\&')}\""
+    end
+  end
+
   @@ruby = rubybin
-  @@good_rake = "#{rubybin} \"#{File.expand_path('../../../test/rubygems/good_rake.rb', __FILE__)}\""
-  @@bad_rake = "#{rubybin} \"#{File.expand_path('../../../test/rubygems/bad_rake.rb', __FILE__)}\""
+  gempath = File.expand_path('../../../test/rubygems', __FILE__)
+  @@good_rake = "#{rubybin} #{escape_path(gempath+'/good_rake.rb')}"
+  @@bad_rake = "#{rubybin} #{escape_path(gempath+'/bad_rake.rb')}"
```
Comment (duckinator) [inline lib/rubygems/test_case.rb]: Could `gempath+'/good_rake.rb'` be replaced with `File.join(gempath, 'good_rake.rb')` (and similar for bad_rake.rb)?

File: lib/rubygems/test_case.rb [hunk 5]
```diff
@@ -1319,9 +1319,18 @@ def self.rubybin
     end
   end
 
+  def self.escape_path(path)
```
Comment (duckinator) [inline lib/rubygems/test_case.rb]: Could you add a comment explaining why `escape_path` is necessary?

**Comment by hsbt** [comment]
@bundlerbot r+

**Pull request merged**